### ISSUE TITLE
Continuous cache invalidation when mill / metals target different JVMs

### ISSUE THREAD

**Issue by Quafadas** [com-lihaoyi/mill#3057]
I observed a situation where mill was continuously invalidating caches. Particularly in the case of scalably typed, this was... frustrating. 

We are fortunate, to have an incredibly knowledgeable community... 

> Are you sure that you are using the same version of JVM in all places?

> Because it happened to me when Metals was running on a JVM and cli on another and Mill was invalidating the cache continuously

I believe, that the problem was exactly this, as after explicitly setting the metals JVM to the same as mill, life got a lot more fun again. 

Metals was auto-detecting an old (v8) JVM via a system config that I can't change. Mill was pointed at the one I configured. 

I'm going to try and setup a reproduction in a dev container - we'll see how it goes. However, without the generosity of @lolgab, I don't see how I would have detected this from a user perspective.  

Given the impact on the user experience, I think it could be worth seeking some sort of mitigation. 

Proposal: when a BSP <-> LSP connection is established, could a JVM root handshake be offered as part of the exchange establishing a connection - with a warning message printed on both sides if it does not pass?

**Comment by Quafadas** [comment]
I think I do mean, exactly this;

> If you mean that the BSP server should use the same JVM as the cli, now, it should. 

Your commentary on the environment variables, is also insightful. I think I'm noticing, that at home, I have no problems, and at $WORK, I think I'm hitting this a lot. 

This is suggestive to me that; 

1. Despite my best efforts to chase down environmental differences, JVMs etc, mill is still being started differently on CLI vs metals. 
2. Windows itself presents some difficulty
3. My corporate environment itself cause a problem (I cannot rule this out - and it would be clearly and obviously not mills problem) 

I've been trying to reproduce this behaviour in a public sandbox (without much joy so far) in an attempt to disentangle 1,2,3, currently without success. 

> Maybe, we could also print an info, when we drop the cache just because of environment changes. This sounds not that hard, but I'm sure, it can be much harder than we expect.

It doesn't sound like this issue is actually super common, as to be honest I would have expected a torrent of prior feedback if it was. So making "hard" changes would not appear justified. I'm wondering in the first instance, if there is an (exhaustive) list of parameters which could cause such behaviour. This could provide a concrete starting point ... 

Could it be possible to gather all necessary information out of startup logging?

**Comment by Quafadas** [comment]
For example; could settings in `.mill-jvm-opts` trigger this?

https://mill-build.com/mill/Builtin_Commands.html#_running_mill_with_custom_jvm_options

**Comment by lefou** [comment]
Maybe we can log the output of `Mill --version` when starting the BSP server...

**Comment by lefou** [comment]
@Quafadas What's in your BSP connection file `.bsp/mill-bsp.json`?

**Comment by lefou** [comment]
Since the BSP client is just starting the executable specified with `argv`, you can replace that with some executable script in which you can inspect and log as much as you want. You should only print to STDERR, as STDOUT is reserved for the build server protocol, though.

**Comment by Quafadas** [comment]
```
{
  "name": "mill-bsp",
  "argv": [
    "C:\\Users\\partens\\scala-cli\\mill.bat",
    "--bsp",
    "--disable-ticker",
    "--color",
    "false",
    "--jobs",
    "1"
  ],
  "millVersion": "0.11.7",
  "bspVersion": "2.2.0-M1",
  "languages": [
    "scala",
    "java"
  ]
}
```
This,
`C:\\Users\\partens\\scala-cli\\mill.bat`

is on the windows path, it is a clone of millw with some light edits (coursier repositories to look for).
![image](https://github.com/com-lihaoyi/mill/assets/24899792/227cd3f5-cae1-4526-96c0-299b6b76f3e6)

I also tried to make sure, that metals used he same script for anything it launched internally.

Finally, I can prove that each call to `mill` from a command line, goes through that script, as echo statements are printed to the console.

This was my naive attempt, to chase down any config differences.

**Comment by lefou** [comment]
Can you run `mill.bat --version` from it? Maybe, you get some insight about the used JVM?

**Comment by Quafadas** [comment]
```
PS C:\temp\Ils-Origination> mill --version
Could not start a Mill server process.
This could be caused by too many already running Mill instances or by an unsupported platform.

Trying to run Mill in-process ...
Launching Mill as sub-process ...
Mill Build Tool version 0.11.7
Java version: 21.0.2, vendor: Microsoft, runtime: C:\temp\jdk-21.0.2+13
Default locale: en_US, platform encoding: UTF-8
OS name: "Windows 10", version: 10.0, arch: amd64
```

**Comment by lefou** [comment]
You can run `rm -r out/mill-worker-*` to fix the first error message.

Make sure the `mill --version` output is identical from your cli and from inside Metals.

**Comment by Quafadas** [comment]
For the lolz, I also did this, 
```
PS C:\temp\Ils-Origination> mill --bsp --disable-ticker
Launching Mill as sub-process ...
Mill in BSP mode, version 0.11.7, Wed Mar 06 11:24:30 CET 2024
Running in BSP mode with hardcoded startSession command
Trying to load BSP server...
Reading worker classpath from file: C:\temp\Ils-Origination\.bsp\mill-bsp-0.11.7.resources
BSP server started
[#0] BSP/startSession: Starting BSP session
```
which looks ... fine?

**Comment by lefou** [comment]
I think so. You could start sending it JSON-RCP messages, which is not fun on the cli, but it should work.

Can you change your `mill.bat` to always print the output of `mill --version` to STDERR? I can't tell you how that works on Windows, though.

**Comment by lefou** [comment]
With bash, I would just add a `1>&2`. 

Put the --version run it in the second last line before the line with `%MILL%`

**Comment by lefou** [comment]
Looks like Metals isn't showing the output in it's logs. Maybe redirecting to a file works better.

**Comment by lefou** [comment]
Oh, I have another idea. When Mill is starting up, it also captures some runner state. Search the `.bsp/mill-bsp.stderr` file for a line starting with "Updateing Evaluator". It should contain all env variables.

**Comment by Quafadas** [comment]
Apologies for disappearing - dragged away for one of the kids. 

> Search the .bsp/mill-bsp.stderr file for a line starting with "Updating Evaluator".

I found this to be an extremely profitable direction of travel. I think it proved that, despite my best efforts another JAVA was in fact making it's way onto the path. It also inspired me to make a couple of other discoveries. 

- Discovered a reference to an old JVM on the PATH. Which I've removed
- Set the `COURISER_CACHE` variable to point to c:\temp. I realised it was directed at my profile, which is happens to be very slow in er... some environment. 

And now it's like that moment in films when the "Beautiful Blue Danube" starts playing in the background. The things I've changed are all local to me, but I think the hint here was very valuable. 

Thankyou, so much for your advice 🙏. I will put up a doc PR in the coming days with it and close this issue.
